feat: add token deployment implementation for EVM and Solana

jkongie opened the pull request:
This commit introduces the implementation for deploying tokens on EVM and Solana networks using the Operations API. These changesets will eventually replace the existing token deployment logic in the `common` package.

By introducing a new `tokens` package, we aim to provide more structure to how changesets related to tokens are handled, and this also serves as an example of how to use ChangeSetV2 and Operations API for building Changesets.

AnieeG commented on deployment/tokens/internal/seqs/seq_persist_address.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// SeqPersistAddressInput is the input to the SeqPersistAddress sequence.
+type SeqPersistAddressInput struct {
```
looks like it should be a generic sequence not specific to tokens. makes sense to move it out to common?

jkongie replied:
I've left this in `internal` for now while we explore how we want to expose common operations through the framework.

The problem with using `common` is that it it becomes a bucket for anyone who doesn't know where to put something in, which makes code ownership much harder. We will be slowly unwinding usage of the `common` pacakge

graham-chainlink reviewed (commented):
nitpick: in terms of package naming, i prefer operation instead of ops and sequence instead of sequences

Because golang packages are conventionally singular, instead of plural and just more obvious instead of guess what op is

graham-chainlink commented on deployment/tokens/internal/ops/ops_evm.go:
```diff
@@ -0,0 +1,91 @@
+	// It is used as part of the unique cache key for the report and in logging.
+	ChainSelector uint64 `json:"chainSelector"`
+	// ChainName is the human friendly name of the chain. This is only used for logging.
+	ChainName string `json:"chainName"`
```
do we need this field, we can use the chainselector library to get the name since we are already passing in the chain selector?

jkongie replied:
I put it in because I didn't want the operation to have to import the `chain_selectors` library just for logging, and deal with the error if it couldn't find it.

Ideally I would just remove it entirely and remove it from the logging, but I left it in for now

graham-chainlink replied:
I see, well with the new chain package, we will have access to the [name and string method ](https://github.com/smartcontractkit/chainlink-deployments-framework/blob/89a82216d2fd5ffb6db8b99efef80bf5b7090434/chain/internal/common/chain_metadata.go#L21-L21) so we not need to do this anymore

jkongie replied:
Updated to use the `ChainInfo` function since we don't want to pass the `Chain` instance down into the operation.

graham-chainlink commented on deployment/tokens/internal/ops/ops_evm.go:
```diff
@@ -0,0 +1,91 @@
+	semver.MustParse("1.0.0"),
+	"Deploy EVM LINK Token Contract",
+	func(b operations.Bundle, deps OpEVMDeployLinkTokenDeps, in OpEVMDeployLinkTokenInput) (OpEvmDeployLinkTokenOutput, error) {
+		out := OpEvmDeployLinkTokenOutput{}
```
I prefer to inline this since this is only used in error case. 

It makes it clear in error case that we are returning an empty struct eg
```
return OpEvmDeployLinkTokenOutput{}, err
```

jkongie replied:
I tend to find this cleaner.

If an error is produced, they shouldn't be using the output anyway

graham-chainlink replied:
yeah, i am okay with this, just comes down to preference. all good!

graham-chainlink commented on deployment/tokens/internal/ops/ops_evm.go:
```diff
@@ -0,0 +1,91 @@
+type OpEVMDeployLinkTokenInput struct {
+	// ChainSelector is the unique identifier for the chain where the operation will be executed.
+	// It is used as part of the unique cache key for the report and in logging.
+	ChainSelector uint64 `json:"chainSelector"`
```
also why we have the `json` tag ?  entirely optional right?

jkongie replied:
Because I want it to unmarshal to lower camel case

graham-chainlink replied:
for operation reports is it?

jkongie replied:
yeah. I guess this is really up to the the conventions each team wants to set. Personally, having uppercase in json fields feels ugly to me

graham-chainlink commented on deployment/tokens/internal/seqs/seq_deploy_evm_tokens.go:
```diff
@@ -0,0 +1,104 @@
+			Addresses: make([]string, 0),
+		}
+
+		for _, csel := range in.ChainSelectors {
```
thinking: i wonder if it is worth using go routine here to parallel deploy the contract or just keep it simple?

jkongie replied:
I'm keeping it simple since I don't really see much of a use case to deploy multiple contracts to multiple chains in a single migration

jkongie commented:
> nitpick: in terms of package naming, i prefer operation instead of ops and sequence instead of sequences
> 
> Because golang packages are conventionally singular, instead of plural and just more obvious instead of guess what op is

So the reason why I chose plural here is because of potential naming conflicts with a variable name vs a package name. I can see someone declaring an `op` var potentially. 

Since this is in the `internal` package, a rename would be localised so it would be good to get this in and while we decide on consistent conventions

simsonraj commented on deployment/tokens/internal/ops/ops_evm.go:
```diff
@@ -0,0 +1,91 @@
+
+	cldf "github.com/smartcontractkit/chainlink-deployments-framework/deployment"
+	"github.com/smartcontractkit/chainlink-deployments-framework/operations"
+	link_token "github.com/smartcontractkit/chainlink-testing-framework/seth/contracts/bind/link"
```
Should we be using Link token from the CTF package & not from chainlink-evm?

jkongie replied:
Good catch! Auto imports fails

jkongie committed:
feat: add token deployment implementation for EVM and Solana

This commit introduces the implementation for deploying tokens on EVM
and Solana networks using the Operations API. These changesets will
eventually replace the existing token deployment logic in the `common`
package.

By introducing a new `tokens` package, we aim to provide more structure
to how changesets related to tokens are handled, and this also serves
as an example of how to use ChangeSetV2 and Operations API for building
Changesets.

graham-chainlink reviewed (commented):
file naming wise for example `ops_evm.go` should be `op_evm.go` similar to the sequence for consistency

graham-chainlink commented on deployment/tokens/internal/ops/ops_evm.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+// OpEVMDeployLinkTokenInput represents the input parameters for the OpEVMDeployLinkToken operation.
+// The chainSelector and chainName fields are used to identify the target chain for deployment
```
`chainName` field doesnt seem to exist anymore?

jkongie replied:
removed the line entirely since it is duplicated in the docstring for the `ChainSelector` field

graham-chainlink commented on deployment/tokens/internal/ops/ops_evm.go:
```diff
@@ -0,0 +1,100 @@
+// deployment logic.
+type OpEVMDeployLinkTokenInput struct {
+	// ChainSelector is the unique identifier for the chain where the operation will be executed.
+	// It is used as part of the unique cache key for the report and in logging.
```
not sure if we should mention this, the cache stuff should be invisible to the users and work transparently without them knowing the details

jkongie replied:
I think it's important to let them know why the chain selector is provided, even though it is technically not used inside the operational logic. 

If we don't have it, someone could remove this because they think it is redundant

graham-chainlink replied:
oh i see now, chainselector is not used in the operation logic

graham-chainlink commented on deployment/tokens/internal/ops/ops_sol.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// OpSolDeployLinkTokenInput represents the input parameters for the OpSolDeployLinkToken operation.
+// The chainSelector and chainName fields are used to identify the target chain for deployment
```
no `chainName` field here too

jkongie replied:
removed the line entirely since it is duplicated in the docstring for the `ChainSelector` field

jkongie commented:
> file naming wise for example `ops_evm.go` should be `op_evm.go` similar to the sequence for consistency

The pluralization was intentional as the plan is to put all the evm related operations in this file, whereas the `seq` files were intented to only contain a single sequence.

graham-chainlink commented on deployment/tokens/changesets/validate.go:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	// Validate there is no existing token contract in the datastore
+	for _, csel := range csels {
```
you could technically combine the 2 for loops here into a single one but i dont mind this if you want to emphasise the separation

jkongie replied:
I kept it separate so that it is easier to remove the address book loop once we fully deprecate the address book

graham-chainlink replied:
yeh i thought so too, looks good then!

graham-chainlink commented on deployment/tokens/changesets/validate_test.go:
```diff
@@ -0,0 +1,188 @@
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := validateNoDupeSelectors(tt.give)
```
t.Parallel()
